### DISCUSSION TITLE
Using prebuilt Windows build in MSVC CMake Project

### DISCUSSION THREAD

**Discussion by rafaelsamenezes** [cvc5/cvc5#10467] [Q&A]
Hello CVC5 team,

I have a CMake project and am trying to link using the prebuilt CVC5 from https://github.com/cvc5/cvc5/releases/tag/cvc5-1.1.2

However, it is failing with an error: `C:\Users\rafae\Downloads\cvc5-Win64-shared\cvc5-Win64-shared\include\cvc5/cvc5_kind.h(5634): error C2144: syntax error: 'int' should be preceded by ';'`. The line contains `[[deprecated("use std::to_string(Kind) instead.")]] std::string kindToString(Kind kind) CVC5_EXPORT;`

From my understanding the CVC5_EXPORT which adds ` __declspec(dllimport)`. I am not really familiar with using DLLs but [MSVC docs](https://learn.microsoft.com/en-us/cpp/cpp/declspec?view=msvc-170) states `The __declspec keywords should be placed at the beginning of a simple declaration. The compiler ignores, without warning, any __declspec keywords placed after * or & and in front of the variable identifier in a declaration.`

I also constructed a mini cmake project which replicates the issue that I am having

```
# CMakeLists.txt
cmake_minimum_required(VERSION 3.27)
project(cvc5_test)

set(CMAKE_CXX_STANDARD 17)

find_package(cvc5 REQUIRED
        HINTS ${CVC5_DIR}/lib/cmake ${CVC5_DIR}/lib64/cmake)

message(STATUS "CVC5 version: ${cvc5_VERSION}")

add_library(cvc5_test library.cpp)
target_link_libraries(cvc5_test cvc5::cvc5)
```

```
// library.cpp
#include <cvc5/cvc5.h>

void hello() {

}
```

Steps to run:

1. Download the prebuilt release of CVC5
2. Run cmake giving `-DCVC5_DIR=<path-to-prebuilt-release>`
3. Build

Do you have any advice on how to procede?

**Comment by daniel-larraz** [comment]
Thank you for reporting this issue! Windows support is still experimental. It seems we are not placing all the CVC5_EXPORT annotations in the right location. I will create an issue from this discussion to fix this.

**Comment by daniel-larraz** [comment]
@rafaelsamenezes . The [issue](https://github.com/cvc5/cvc5/issues/10475) has been fixed in the main branch.

**Comment by rafaelsamenezes** [comment]
Thanks @daniel-larraz , that solved my initial issue. However, while trying to integrate into ESBMC I am having trouble at the linking stage:

```
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: bool __cdecl cvc5::Result::isSat(void)const " (__imp_?isSat@Result@cvc5@@QEBA_NXZ) referenced in function "public: virtual enum smt_convt::resultt __cdecl cvc_convt::dec_solve(void)" (?dec_solve@cvc_convt@@UEAA?AW4resultt@smt_convt@@XZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: bool __cdecl cvc5::Result::isUnknown(void)const " (__imp_?isUnknown@Result@cvc5@@QEBA_NXZ) referenced in function "public: virtual enum smt_convt::resultt __cdecl cvc_convt::dec_solve(void)" (?dec_solve@cvc_convt@@UEAA?AW4resultt@smt_convt@@XZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl cvc5::Result::~Result(void)" (__imp_??1Result@cvc5@@QEAA@XZ) referenced in function "public: virtual enum smt_convt::resultt __cdecl cvc_convt::dec_solve(void)" (?dec_solve@cvc_convt@@UEAA?AW4resultt@smt_convt@@XZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl cvc5::Sort::~Sort(void)" (__imp_??1Sort@cvc5@@QEAA@XZ) referenced in function "public: virtual __cdecl solver_smt_sort<class cvc5::Sort>::~solver_smt_sort<class cvc5::Sort>(void)" (??1?$solver_smt_sort@VSort@cvc5@@@@UEAA@XZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl cvc5::Sort::Sort(class cvc5::Sort const &)" (__imp_??0Sort@cvc5@@QEAA@AEBV01@@Z) referenced in function "public: virtual class smt_sort const * __cdecl cvc_convt::mk_array_sort(class smt_sort const *,class smt_sort const *)" (?mk_array_sort@cvc_convt@@UEAAPEBVsmt_sort@@PEBV2@0@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl cvc5::Op::~Op(void)" (__imp_??1Op@cvc5@@QEAA@XZ) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_extract(class smt_ast const *,unsigned int,unsigned int)" (?mk_extract@cvc_convt@@UEAAPEBVsmt_ast@@PEBV2@II@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl cvc5::Term::~Term(void)" (__imp_??1Term@cvc5@@QEAA@XZ) referenced in function "public: class smt_ast const * __cdecl smt_convt::new_solver_ast<class cvc_smt_ast>(class cvc5::Term,class smt_sort const *)" (??$new_solver_ast@Vcvc_smt_ast@@@smt_convt@@QEAAPEBVsmt_ast@@VTerm@cvc5@@PEBVsmt_sort@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: enum cvc5::Kind __cdecl cvc5::Term::getKind(void)const " (__imp_?getKind@Term@cvc5@@QEBA?AW4Kind@2@XZ) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_from_fp_to_bv(class smt_ast const *)" (?mk_from_fp_to_bv@cvc_convt@@UEAAPEBVsmt_ast@@PEBV2@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl cvc5::Term::toString(void)const " (__imp_?toString@Term@cvc5@@QEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) referenced in function "public: virtual void __cdecl cvc_smt_ast::dump(void)const " (?dump@cvc_smt_ast@@UEBAXXZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: bool __cdecl cvc5::Term::getBooleanValue(void)const " (__imp_?getBooleanValue@Term@cvc5@@QEBA_NXZ) referenced in function "public: virtual bool __cdecl cvc_convt::get_bool(class smt_ast const *)" (?get_bool@cvc_convt@@UEAA_NPEBVsmt_ast@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl cvc5::Term::getBitVectorValue(unsigned int)const " (__imp_?getBitVectorValue@Term@cvc5@@QEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@I@Z) referenced in function "public: virtual class BigInt __cdecl cvc_convt::get_bv(class smt_ast const *,bool)" (?get_bv@cvc_convt@@UEAA?AVBigInt@@PEBVsmt_ast@@_N@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: bool __cdecl cvc5::Term::isFloatingPointNaN(void)const " (__imp_?isFloatingPointNaN@Term@cvc5@@QEBA_NXZ) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_from_fp_to_bv(class smt_ast const *)" (?mk_from_fp_to_bv@cvc_convt@@UEAAPEBVsmt_ast@@PEBV2@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl cvc5::Term::Term(class cvc5::Term const &)" (__imp_??0Term@cvc5@@QEAA@AEBV01@@Z) referenced in function "public: class smt_ast const * __cdecl smt_convt::new_solver_ast<class cvc_smt_ast>(class cvc5::Term,class smt_sort const *)" (??$new_solver_ast@Vcvc_smt_ast@@@smt_convt@@QEAAPEBVsmt_ast@@VTerm@cvc5@@PEBVsmt_sort@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl cvc5::Solver::Solver(void)" (__imp_??0Solver@cvc5@@QEAA@XZ) referenced in function "public: __cdecl cvc_convt::cvc_convt(class namespacet const &,class optionst const &)" (??0cvc_convt@@QEAA@AEBVnamespacet@@AEBVoptionst@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl cvc5::Solver::~Solver(void)" (__imp_??1Solver@cvc5@@QEAA@XZ) referenced in function "int `public: __cdecl cvc_convt::cvc_convt(class namespacet const &,class optionst const &)'::`1'::dtor$1" (?dtor$1@?0???0cvc_convt@@QEAA@AEBVnamespacet@@AEBVoptionst@@@Z@4HA) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Sort __cdecl cvc5::Solver::getBooleanSort(void)const " (__imp_?getBooleanSort@Solver@cvc5@@QEBA?AVSort@2@XZ) referenced in function "public: virtual class smt_sort const * __cdecl cvc_convt::mk_bool_sort(void)" (?mk_bool_sort@cvc_convt@@UEAAPEBVsmt_sort@@XZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Sort __cdecl cvc5::Solver::getIntegerSort(void)const " (__imp_?getIntegerSort@Solver@cvc5@@QEBA?AVSort@2@XZ) referenced in function "public: virtual class smt_sort const * __cdecl cvc_convt::mk_int_sort(void)" (?mk_int_sort@cvc_convt@@UEAAPEBVsmt_sort@@XZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Sort __cdecl cvc5::Solver::getRealSort(void)const " (__imp_?getRealSort@Solver@cvc5@@QEBA?AVSort@2@XZ) referenced in function "public: virtual class smt_sort const * __cdecl cvc_convt::mk_real_sort(void)" (?mk_real_sort@cvc_convt@@UEAAPEBVsmt_sort@@XZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Sort __cdecl cvc5::Solver::getRoundingModeSort(void)const " (__imp_?getRoundingModeSort@Solver@cvc5@@QEBA?AVSort@2@XZ) referenced in function "public: virtual class smt_sort const * __cdecl cvc_convt::mk_fpbv_rm_sort(void)" (?mk_fpbv_rm_sort@cvc_convt@@UEAAPEBVsmt_sort@@XZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Sort __cdecl cvc5::Solver::mkArraySort(class cvc5::Sort const &,class cvc5::Sort const &)const " (__imp_?mkArraySort@Solver@cvc5@@QEBA?AVSort@2@AEBV32@0@Z) referenced in function "public: virtual class smt_sort const * __cdecl cvc_convt::mk_array_sort(class smt_sort const *,class smt_sort const *)" (?mk_array_sort@cvc_convt@@UEAAPEBVsmt_sort@@PEBV2@0@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Sort __cdecl cvc5::Solver::mkBitVectorSort(unsigned int)const " (__imp_?mkBitVectorSort@Solver@cvc5@@QEBA?AVSort@2@I@Z) referenced in function "public: virtual class smt_sort const * __cdecl cvc_convt::mk_bv_sort(unsigned __int64)" (?mk_bv_sort@cvc_convt@@UEAAPEBVsmt_sort@@_K@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Sort __cdecl cvc5::Solver::mkFloatingPointSort(unsigned int,unsigned int)const " (__imp_?mkFloatingPointSort@Solver@cvc5@@QEBA?AVSort@2@II@Z) referenced in function "public: virtual class smt_sort const * __cdecl cvc_convt::mk_fpbv_sort(unsigned int,unsigned int)" (?mk_fpbv_sort@cvc_convt@@UEAAPEBVsmt_sort@@II@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkTerm(enum cvc5::Kind,class std::vector<class cvc5::Term,class std::allocator<class cvc5::Term> > const &)const " (__imp_?mkTerm@Solver@cvc5@@QEBA?AVTerm@2@W4Kind@2@AEBV?$vector@VTerm@cvc5@@V?$allocator@VTerm@cvc5@@@std@@@std@@@Z) referenced in function "public: virtual class irep_container<class expr2t> __cdecl cvc_convt::get_array_elem(class smt_ast const *,unsigned __int64,class irep_container<class type2t> const &)" (?get_array_elem@cvc_convt@@UEAA?AV?$irep_container@Vexpr2t@@@@PEBVsmt_ast@@_KAEBV?$irep_container@Vtype2t@@@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkTerm(class cvc5::Op const &,class std::vector<class cvc5::Term,class std::allocator<class cvc5::Term> > const &)const " (__imp_?mkTerm@Solver@cvc5@@QEBA?AVTerm@2@AEBVOp@2@AEBV?$vector@VTerm@cvc5@@V?$allocator@VTerm@cvc5@@@std@@@std@@@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_extract(class smt_ast const *,unsigned int,unsigned int)" (?mk_extract@cvc_convt@@UEAAPEBVsmt_ast@@PEBV2@II@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Op __cdecl cvc5::Solver::mkOp(enum cvc5::Kind,class std::initializer_list<unsigned int> const &)const " (__imp_?mkOp@Solver@cvc5@@QEBA?AVOp@2@W4Kind@2@AEBV?$initializer_list@I@std@@@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_extract(class smt_ast const *,unsigned int,unsigned int)" (?mk_extract@cvc_convt@@UEAAPEBVsmt_ast@@PEBV2@II@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkBoolean(bool)const " (__imp_?mkBoolean@Solver@cvc5@@QEBA?AVTerm@2@_N@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_smt_bool(bool)" (?mk_smt_bool@cvc_convt@@UEAAPEBVsmt_ast@@_N@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkInteger(__int64)const " (__imp_?mkInteger@Solver@cvc5@@QEBA?AVTerm@2@_J@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_smt_int(class BigInt const &)" (?mk_smt_int@cvc_convt@@UEAAPEBVsmt_ast@@AEBVBigInt@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkReal(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)const " (__imp_?mkReal@Solver@cvc5@@QEBA?AVTerm@2@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_smt_real(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)" (?mk_smt_real@cvc_convt@@UEAAPEBVsmt_ast@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkBitVector(unsigned int,unsigned __int64)const " (__imp_?mkBitVector@Solver@cvc5@@QEBA?AVTerm@2@I_K@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_smt_bv(class BigInt const &,class smt_sort const *)" (?mk_smt_bv@cvc_convt@@UEAAPEBVsmt_ast@@AEBVBigInt@@PEBVsmt_sort@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkFloatingPointPosInf(unsigned int,unsigned int)const " (__imp_?mkFloatingPointPosInf@Solver@cvc5@@QEBA?AVTerm@2@II@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_smt_fpbv_inf(bool,unsigned int,unsigned int)" (?mk_smt_fpbv_inf@cvc_convt@@UEAAPEBVsmt_ast@@_NII@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkFloatingPointNegInf(unsigned int,unsigned int)const " (__imp_?mkFloatingPointNegInf@Solver@cvc5@@QEBA?AVTerm@2@II@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_smt_fpbv_inf(bool,unsigned int,unsigned int)" (?mk_smt_fpbv_inf@cvc_convt@@UEAAPEBVsmt_ast@@_NII@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkFloatingPointNaN(unsigned int,unsigned int)const " (__imp_?mkFloatingPointNaN@Solver@cvc5@@QEBA?AVTerm@2@II@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_smt_fpbv_nan(bool,unsigned int,unsigned int)" (?mk_smt_fpbv_nan@cvc_convt@@UEAAPEBVsmt_ast@@_NII@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkRoundingMode(enum cvc5::RoundingMode)const " (__imp_?mkRoundingMode@Solver@cvc5@@QEBA?AVTerm@2@W4RoundingMode@2@@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_smt_fpbv_rm(enum ieee_floatt::rounding_modet)" (?mk_smt_fpbv_rm@cvc_convt@@UEAAPEBVsmt_ast@@W4rounding_modet@ieee_floatt@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkFloatingPoint(unsigned int,unsigned int,class cvc5::Term const &)const " (__imp_?mkFloatingPoint@Solver@cvc5@@QEBA?AVTerm@2@IIAEBV32@@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_smt_fpbv(class ieee_floatt const &)" (?mk_smt_fpbv@cvc_convt@@UEAAPEBVsmt_ast@@AEBVieee_floatt@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::mkConst(class cvc5::Sort const &,class std::optional<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > const &)const " (__imp_?mkConst@Solver@cvc5@@QEBA?AVTerm@2@AEBVSort@2@AEBV?$optional@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@std@@@Z) referenced in function "public: virtual class smt_ast const * __cdecl cvc_convt::mk_smt_symbol(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class smt_sort const *)" (?mk_smt_symbol@cvc_convt@@UEAAPEBVsmt_ast@@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@PEBVsmt_sort@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: void __cdecl cvc5::Solver::assertFormula(class cvc5::Term const &)const " (__imp_?assertFormula@Solver@cvc5@@QEBAXAEBVTerm@2@@Z) referenced in function "public: virtual void __cdecl cvc_convt::assert_ast(class smt_ast const *)" (?assert_ast@cvc_convt@@UEAAXPEBVsmt_ast@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Result __cdecl cvc5::Solver::checkSat(void)const " (__imp_?checkSat@Solver@cvc5@@QEBA?AVResult@2@XZ) referenced in function "public: virtual enum smt_convt::resultt __cdecl cvc_convt::dec_solve(void)" (?dec_solve@cvc_convt@@UEAA?AW4resultt@smt_convt@@XZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::vector<class cvc5::Term,class std::allocator<class cvc5::Term> > __cdecl cvc5::Solver::getAssertions(void)const " (__imp_?getAssertions@Solver@cvc5@@QEBA?AV?$vector@VTerm@cvc5@@V?$allocator@VTerm@cvc5@@@std@@@std@@XZ) referenced in function "public: virtual void __cdecl cvc_convt::dump_smt(void)" (?dump_smt@cvc_convt@@UEAAXXZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class cvc5::Term __cdecl cvc5::Solver::getValue(class cvc5::Term const &)const " (__imp_?getValue@Solver@cvc5@@QEBA?AVTerm@2@AEBV32@@Z) referenced in function "public: virtual bool __cdecl cvc_convt::get_bool(class smt_ast const *)" (?get_bool@cvc_convt@@UEAA_NPEBVsmt_ast@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: void __cdecl cvc5::Solver::setOption(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &)const " (__imp_?setOption@Solver@cvc5@@QEBAXAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@0@Z) referenced in function "public: __cdecl cvc_convt::cvc_convt(class namespacet const &,class optionst const &)" (??0cvc_convt@@QEAA@AEBVnamespacet@@AEBVoptionst@@@Z) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl cvc5::Solver::getVersion(void)const " (__imp_?getVersion@Solver@cvc5@@QEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) referenced in function "public: virtual class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const __cdecl cvc_convt::solver_text(void)" (?solver_text@cvc_convt@@UEAA?BV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
solvercvc5.lib(cvc_conv.obj) : error LNK2001: unresolved external symbol "public: __cdecl cvc5::Result::~Result(void)" (??1Result@cvc5@@QEAA@XZ) [C:\Users\rafae\GitHub\esbmc\cmake-build-relwithdebinfo\src\esbmc\esbmc.vcxproj]
```

**Reply by aytey** [reply]
You probably need a `.def` file to make this work adequately. Likely this is a name mangling failing between what MinGW generates and MSVC wants. See, e.g., https://stackoverflow.com/a/3031167

However, I'd be worried what happens after you resolve that: you're going to have a DLL linking against the GCC C++ runtime and an application (using that DLL) that wants to use the MSVCRT.

I just checked `libcvc5.dll`:

```
Dump of file libcvc5.dll

File Type: DLL

  Image has the following dependencies:

    libgmp-10.dll
    KERNEL32.dll
    msvcrt.dll
    PSAPI.DLL
    libgcc_s_seh-1.dll
    libstdc++-6.dll
    libpoly.dll
    libpolyxx.dll
```

So this has a runtime dependency on `libgcc_s` and `libstdc++`, which won't be fulfilled by MSVCRT.

You *might* be able statically link `libgcc_s` and `libstdc++` into `libcvc5.dll`, but then you're going to have a problem if you pass C++ objects "across the ABI boundary".

Basically, I think you could be setting yourself up for some headaches later on even if this works -- maybe if you *only* used the C API and statically linked the GCC runtimes, you might be okay.
